solidity/contracts/converter/BancorConverter.sol

```diff
@@ -789,6 +782,174 @@ contract BancorConverter is IBancorConverter, TokenHandler, SmartTokenController
             safeTransferFrom(_token, _from, _to, _amount);
     }
 
+    function addLiquidity(IERC20Token[] memory _reserveTokens, uint256[] memory _reserveAmounts, uint256 _supplyMinReturnAmount)
+        public
+        payable
+        multipleReservesOnly
+    {
+        verifyLiquidityInput(_reserveTokens, _reserveAmounts, _supplyMinReturnAmount);
+
+        for (uint256 i = 0; i < _reserveTokens.length; i++)
+            if (_reserveTokens[i] == address(0))
+                require(_reserveAmounts[i] == msg.value);
+
+        if (msg.value > 0)
+            require(reserves[address(0)].isSet);
+
+        uint256 totalSupply = token.totalSupply();
+        uint256 supplyAmount;
+
+        if (totalSupply == 0)
+            supplyAmount = addLiquidityToEmptyPool(_reserveTokens, _reserveAmounts);
+        else
+            supplyAmount = addLiquidityToNonEmptyPool(_reserveTokens, _reserveAmounts, totalSupply);
+
+        require(supplyAmount >= _supplyMinReturnAmount);
+        token.issue(msg.sender, supplyAmount);
+    }
+
+    function removeLiquidity(IERC20Token[] memory _reserveTokens, uint256[] memory _reserveMinReturnAmounts, uint256 _supplyAmount)
+        public
+        multipleReservesOnly
+    {
+        verifyLiquidityInput(_reserveTokens, _reserveMinReturnAmounts, _supplyAmount);
+
+        uint256 totalSupply = token.totalSupply();
+
+        token.destroy(msg.sender, _supplyAmount);
+
+        removeLiquidityFromPool(_reserveTokens, _reserveMinReturnAmounts, totalSupply, _supplyAmount);
+    }
+
+    function verifyLiquidityInput(IERC20Token[] memory _reserveTokens, uint256[] memory _reserveAmounts, uint256 _supplyAmount)
+        private
+        view
+    {
+        uint256 i;
+        uint256 j;
+
+        uint256 length = reserveTokens.length;
+        require(length == _reserveTokens.length);
+        require(length == _reserveAmounts.length);
+
+        for (i = 0; i < length; i++) {
+            require(reserves[_reserveTokens[i]].isSet);
+            for (j = 0; j < length; j++) {
+                if (reserveTokens[i] == _reserveTokens[j])
+                    break;
+            }
+            require(j < length);

```

Comment:
is there any way to actually fail here given that we test the length before?

Reply:
@yudilevi : We need to verify:
1. That the user has provided only the reserve tokens of the converter
2. That the user has provided all the reserve tokens of the converter
The `require` before the `for j` loop verifies #1.
The `require` after the `for j` loop verifies #2.

Reply:
ah yeah I see it now, sababa